### ISSUE TITLE
systemd-resolved: TCP connection is prematurely closed when multiple requests are sent on same connection

### ISSUE THREAD

**Issue by dkfitbit** [systemd/systemd#11332]
**systemd version the issue has been seen with**
239

**Used distribution**
Ubuntu 18.10 "Cosmic"

**Expected behaviour you didn't see**
N DNS queries sent to 53/tcp on 127.0.0.53 should result in N DNS replies.  In particular, `libnss_dns.so` expects to be able to send both an A query and an AAAA query on the same TCP connection.

**Unexpected behaviour you saw**
systemd-resolved immediately closes the TCP connection after sending the first reply.  `libnss_dns.so` treats this as failure of the entire resolution, causing programs like `curl(1)` to fail with "Could not resolve host" if `libnss_dns.so` decides to use TCP.

**Steps to reproduce the problem**
1. Enable binding to TCP
2. Send two queries on one TCP connection, like so:

```bash
root@00938c02ae2d:~# hd two-queries 
00000000  00 2f a7 a1 01 00 00 01  00 00 00 00 00 00 09 61  |./.............a|
00000010  72 74 69 66 61 63 74 73  08 73 69 74 65 2d 6f 70  |rtifacts.site-op|
00000020  73 06 66 69 74 62 69 74  03 63 6f 6d 00 00 01 00  |s.fitbit.com....|
00000030  01 00 2f 91 a9 01 00 00  01 00 00 00 00 00 00 09  |../.............|
00000040  61 72 74 69 66 61 63 74  73 08 73 69 74 65 2d 6f  |artifacts.site-o|
00000050  70 73 06 66 69 74 62 69  74 03 63 6f 6d 00 00 1c  |ps.fitbit.com...|
00000060  00 01                                             |..|
00000062
root@00938c02ae2d:~# nc 127.0.0.53 53 < two-queries | hd
00000000  01 fb a7 a1 83 80 00 01  00 1a 00 00 00 00 09 61  |...............a|
00000010  72 74 69 66 61 63 74 73  08 73 69 74 65 2d 6f 70  |rtifacts.site-op|
00000020  73 06 66 69 74 62 69 74  03 63 6f 6d 00 00 01 00  |s.fitbit.com....|
00000030  01 c0 0c 00 05 00 01 00  00 00 3b 00 30 18 68 61  |..........;.0.ha|
00000040  70 72 6f 78 79 2d 70 72  69 76 61 74 65 2d 69 6e  |proxy-private-in|
00000050  74 65 72 6e 61 6c 04 70  72 6f 64 05 64 61 6c 30  |ternal.prod.dal0|
00000060  39 09 64 69 73 63 6f 76  65 72 79 c0 1f c0 3b 00  |9.discovery...;.|
00000070  01 00 01 00 00 00 01 00  04 0a 99 7b 5c c0 3b 00  |...........{\.;.|
00000080  01 00 01 00 00 00 01 00  04 0a 99 0b eb c0 3b 00  |..............;.|
00000090  01 00 01 00 00 00 01 00  04 0a 99 7b 3a c0 3b 00  |...........{:.;.|
000000a0  01 00 01 00 00 00 01 00  04 0a 63 5c df c0 3b 00  |..........c\..;.|
000000b0  01 00 01 00 00 00 01 00  04 0a 99 7b 96 c0 3b 00  |...........{..;.|
000000c0  01 00 01 00 00 00 01 00  04 0a 99 7b ad c0 3b 00  |...........{..;.|
000000d0  01 00 01 00 00 00 01 00  04 0a 99 7b 89 c0 3b 00  |...........{..;.|
000000e0  01 00 01 00 00 00 01 00  04 0a 99 7b 65 c0 3b 00  |...........{e.;.|
000000f0  01 00 01 00 00 00 01 00  04 0a 99 0a 18 c0 3b 00  |..............;.|
00000100  01 00 01 00 00 00 01 00  04 0a 99 7b 09 c0 3b 00  |...........{..;.|
00000110  01 00 01 00 00 00 01 00  04 0a 99 0c a9 c0 3b 00  |..............;.|
00000120  01 00 01 00 00 00 01 00  04 0a 99 7b 21 c0 3b 00  |...........{!.;.|
00000130  01 00 01 00 00 00 01 00  04 0a 99 7b b6 c0 3b 00  |...........{..;.|
00000140  01 00 01 00 00 00 01 00  04 0a 99 02 09 c0 3b 00  |..............;.|
00000150  01 00 01 00 00 00 01 00  04 0a 99 7b 8a c0 3b 00  |...........{..;.|
00000160  01 00 01 00 00 00 01 00  04 0a 99 00 90 c0 3b 00  |..............;.|
00000170  01 00 01 00 00 00 01 00  04 0a 99 0b 51 c0 3b 00  |............Q.;.|
00000180  01 00 01 00 00 00 01 00  04 0a 99 09 86 c0 3b 00  |..............;.|
00000190  01 00 01 00 00 00 01 00  04 0a 99 06 66 c0 3b 00  |............f.;.|
000001a0  01 00 01 00 00 00 01 00  04 0a 63 5c 69 c0 3b 00  |..........c\i.;.|
000001b0  01 00 01 00 00 00 01 00  04 0a 99 00 38 c0 3b 00  |............8.;.|
000001c0  01 00 01 00 00 00 01 00  04 0a 99 04 56 c0 3b 00  |............V.;.|
000001d0  01 00 01 00 00 00 01 00  04 0a 99 7b 90 c0 3b 00  |...........{..;.|
000001e0  01 00 01 00 00 00 01 00  04 0a 99 08 f6 c0 3b 00  |..............;.|
000001f0  01 00 01 00 00 00 01 00  04 0a 99 00 e1           |.............|
000001fd
```

[two-queries.gz](https://github.com/systemd/systemd/files/2728433/two-queries.gz)

**Comment by ddstreet** [comment]
quite a coincidence, I was about to open a systemd bug for this as I am working on this also and have opened the Ubuntu bug 1811471:
https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1811471

@poettering I was planning to add DNS pipelining support to the stub resolver, which I think is really the only way to correctly fix this bug; please let me know if you disagree with that direction (so I don't waste my time), or if you already have some fix you're working on.

**Comment by poettering** [comment]
@ddstreet  what do you mean by pipelining? for the stub or for resolved's own communication with the upstream DNS servers? The latter already uses  long-living TCP connection if possible. It was introduced as part of the DNS-over-TLS support, as that's the only way to make that efficient. Or what do you mean?

**Comment by ddstreet** [comment]
I only meant for the stub resolver; the problem is that glibc uses "pipelining" (as described in the rfc https://tools.ietf.org/html/rfc7766#section-6.2.1.1 but basically just multiple requests using a single TCP connection, yes) but the stub resolver doesn't handle that; it services only a single incoming DNS query and then closes the connection.  I do have more technical details in the Ubuntu bug linked above.

**Comment by poettering** [comment]
@ddstreet yupp we really should implement pipelining i guess. shouldn't be too hard, given that we do it in the other direction already (i.e. when systemd is client to some other DNS peer) and the stream implementation is shared between both directions.

(will bump this to v242 though, since this is highly relevant but not a regression)

**Comment by poettering** [comment]
Proposed fix pending in #11512. Any chance anyone can given this a whirl? Would be great to have some testing feedback there! Thanks!
